images/virtualization-artifact/pkg/controller/vd/vd_reconciler.go

```diff
@@ -0,0 +1,209 @@
+		newRequeue, err = h.Handle(ctx, vd.Changed())
+		if err != nil {
+			slog.Error("Failed to handle vd", "err", err, "handler", reflect.TypeOf(h).Elem().Name())
+			handlerErrs = append(handlerErrs, err)
+		}
+
+		requeue = requeue || newRequeue
+	}
+
+	vd.Changed().Status.ObservedGeneration = vd.Changed().Generation
+
+	slog.Info("Update")
+
+	err = vd.Update(ctx)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	err = errors.Join(handlerErrs...)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if requeue {
+		slog.Info("Requeue")
+		return reconcile.Result{
+			RequeueAfter: 5 * time.Second,
+		}, nil
+	}
+
+	slog.Info("Done")
+
+	return reconcile.Result{}, nil
+}
+
+func (r *Reconciler) SetupController(_ context.Context, mgr manager.Manager, ctr controller.Controller) error {
+	if err := ctr.Watch(
+		source.Kind(mgr.GetCache(), &virtv2.VirtualDisk{}),
+		&handler.EnqueueRequestForObject{},
+		predicate.Funcs{
+			CreateFunc: func(e event.CreateEvent) bool { return true },
+			DeleteFunc: func(e event.DeleteEvent) bool { return true },
+			UpdateFunc: func(e event.UpdateEvent) bool {
+				return e.ObjectOld.GetGeneration() != e.ObjectNew.GetGeneration()
+			},
+		},
+	); err != nil {
+		return fmt.Errorf("error setting watch on VirtualDisk: %w", err)
+	}
+
+	if err := ctr.Watch(
+		source.Kind(mgr.GetCache(), &cdiv1.DataVolume{}),
+		handler.EnqueueRequestForOwner(
+			mgr.GetScheme(),
+			mgr.GetRESTMapper(),
+			&virtv2.VirtualDisk{},
+			handler.OnlyControllerOwner(),
+		),
+	); err != nil {
+		return fmt.Errorf("error setting watch on DV: %w", err)
+	}
+
+	if err := ctr.Watch(
+		source.Kind(mgr.GetCache(), &corev1.PersistentVolumeClaim{}),
+		handler.EnqueueRequestForOwner(
+			mgr.GetScheme(),
+			mgr.GetRESTMapper(),
+			&virtv2.VirtualDisk{},
+			handler.OnlyControllerOwner(),
+		), predicate.Funcs{
+			CreateFunc: func(e event.CreateEvent) bool { return false },
+			DeleteFunc: func(e event.DeleteEvent) bool { return true },
+			UpdateFunc: func(e event.UpdateEvent) bool { return false },
+		},
+	); err != nil {
+		return fmt.Errorf("error setting watch on PVC: %w", err)
+	}
+
+	if err := ctr.Watch(
+		source.Kind(mgr.GetCache(), &virtv2.VirtualMachine{}),
+		handler.EnqueueRequestsFromMapFunc(r.enqueueRequests()),
+		predicate.Funcs{
+			CreateFunc: func(e event.CreateEvent) bool {
+				return r.filterAttachedObj(e.Object)
+			},
+			DeleteFunc: func(e event.DeleteEvent) bool {
+				return r.filterAttachedObj(e.Object)
+			},
+			UpdateFunc: func(e event.UpdateEvent) bool {
+				return r.filterAttachedObj(e.ObjectOld) || r.filterAttachedObj(e.ObjectNew)
+			},
+		},
+	); err != nil {
+		return fmt.Errorf("error setting watch on VMs: %w", err)
+	}
+
+	return nil
+}
+
+func (r *Reconciler) factory() *virtv2.VirtualDisk {
+	return &virtv2.VirtualDisk{}
+}
+
+func (r *Reconciler) statusGetter(obj *virtv2.VirtualDisk) virtv2.VirtualDiskStatus {
+	return obj.Status
+}
```

Comment:
Are those needed? Some temporary methods?

Reply:
We still need them